### ISSUE TITLE
[Datasets] [Feature] Add datasource for canned ML datasets and imagery data sources

### ISSUE THREAD

**Issue by clarkzinzow** [ray-project/ray#23977]
Being able to quickly work with familiar canned datasets exposed by ML frameworks (e.g. CIFAR10, MNIST, ImageNet, etc.) is great for a "getting started" flow, since the user doesn't have to worry about finding a dataset to try out and the user is familiar with these common ML datasets. We should add a Datasets datasource for some of these canned datasets to improve our "getting started" UX.

### Requirements

1. [P0] Expose a generic Torch dataset wrapper that downloads the entire dataset onto a single node and builds a single block. This should provide an easy "getting started" path for smaller canned datasets.
2. [P0] Do the same for TensorFlow datasets.
3. [P0] Validate two common image loading workflows for larger-scale (multi-node, sharded) datasets:
  a. **Image-only dataset:** Take a list of image URLs, map a download function over them that (i) parses the class from the path, and (ii) downloads the image. To test this, add one or two of the larger canned datasets (e.g. ImageNet is multiple TB) to a public S3 bucket (or find a large one that’s already hosted). `ray.data.from_items(s3_image_urls).map(download_fn)`
  b. **Images-in-tables dataset:** Read a tabular (Parquet, CSV) dataset that contains an image URL column, map a download function over the blocks that takes the image URL for each row and adds the image to a new column. `ray.data.read_parquet(tabular_ds_url).map_batches(...)`
4. [P1] If 3(a) proves to be bad UX, create a custom imagery-specific datasource that reads the data in parallel from said bucket. This should demonstrate some of the scalable parallel I/O benefits of Datasets and will be a nice “how would I scale this to larger datasets” progression.

**Comment by dynamicwebpaige** [comment]
Love it! Could we also add doing the same for [Hugging Face Datasets](https://huggingface.co/datasets) (roughly identical to TensorFlow Datasets)? 😄

**Comment by clarkzinzow** [comment]
@dynamicwebpaige Since this issue is very focused on canned image datasets and imagery data sources, I'll open a separate issue for doing the same for HuggingFace datasets!

**Comment by clarkzinzow** [comment]
@dynamicwebpaige Created two issues:
- https://github.com/ray-project/ray/issues/24046
- https://github.com/ray-project/ray/issues/24045
The second needs more scoping, possibly after a PoC and/or talking to the HuggingFace datasets folks.

**Comment by clarkzinzow** [comment]
@zhe-thoughts The P0s are, the P1 is not. The only remaining P0, PR [here](https://github.com/ray-project/ray/pull/24022), should be merged by EOD.

**Comment by clarkzinzow** [comment]
With (1), (2), and (3) done, downgrading the remaining work, (4), to a P2.
